algorithms/scripts/RIT1-PCA-htcai.py

```diff
@@ -0,0 +1,257 @@
+
+
+# ## Parameters and Classifier Fitting
+
+# In[15]:
+
+param_grid = {
+    'alpha': [2**x for x in range(-20, 30)],
+    'l1_ratio': [0, 0.05, 0.1, 0.15]
+}
+
+
+# In[16]:
+
+sss = StratifiedShuffleSplit(n_splits=100, test_size=0.1, random_state=0)
+clf = SGDClassifier(random_state=0, class_weight='balanced', loss='log', penalty='elasticnet')
+cv = GridSearchCV(estimator=clf, param_grid=param_grid, n_jobs=-1, scoring='roc_auc', cv=sss)
+
+
+# In[17]:
+
+get_ipython().run_cell_magic('time', '', 'cv.fit(X = X_train_scale, y=y_train)')
+
+
+# In[18]:
+
+# Best Params
+print('{:.3%}'.format(cv.best_score_))
+
+# Best Params
+cv.best_params_
+
+
+# ## Visualize hyperparameters performance
+
+# In[19]:
+
+cv_result_df = pd.concat([
+    pd.DataFrame(cv.cv_results_),
+    pd.DataFrame.from_records(cv.cv_results_['params']),
+], axis='columns')
+cv_result_df.head(2)
+
+
+# In[20]:
+
+# Cross-validated performance heatmap
+cv_score_mat = pd.pivot_table(cv_result_df, values='mean_test_score', index='l1_ratio', columns='alpha')
+fig, ax = plt.subplots(figsize=(15,2))
+
+xticks = ['2^'+str(x) for x in range(-20, 30)]
+keptticks = xticks[::int(len(xticks)/10)]
+xticks = ['' for y in xticks]
+xticks[::int(len(xticks)/10)] = keptticks
+
+ax = sns.heatmap(cv_score_mat, annot=False, fmt='.1%', xticklabels=xticks)
+ax.set_xlabel('Regularization strength multiplier (alpha)')
+ax.set_ylabel('Elastic net mixing parameter (l1_ratio)');
+
+
+# ## Coefficients of the Classifier
+
+# In[21]:
+
+best_clf = cv.best_estimator_
+coef = best_clf.coef_[0]
+plt.figure(figsize = (15, 5))
+colors = ["red" if coef[i] < 0 else "blue" for i in range(len(coef))]
+plt.bar(np.arange(len(coef)), coef, color = colors)
+plt.xticks(np.arange(1, len(coef)+1), rotation=45, ha="right");
+
+
+# ## Use Optimal Hyperparameters to Output ROC Curve
+
+# In[22]:
+
+y_pred_train = cv.decision_function(X_train_scale)
+y_pred_test = cv.decision_function(X_test_scale)
+
+def get_threshold_metrics(y_true, y_pred):
+    roc_columns = ['fpr', 'tpr', 'threshold']
+    roc_items = zip(roc_columns, roc_curve(y_true, y_pred))
+    roc_df = pd.DataFrame.from_items(roc_items)
+    auroc = roc_auc_score(y_true, y_pred)
+    return {'auroc': auroc, 'roc_df': roc_df}
+
+metrics_train = get_threshold_metrics(y_train, y_pred_train)
+metrics_test = get_threshold_metrics(y_test, y_pred_test)
+
+
+# In[23]:
+
+# Plot ROC
+plt.figure()
+for label, metrics in ('Training', metrics_train), ('Testing', metrics_test):
+    roc_df = metrics['roc_df']
+    plt.plot(roc_df.fpr, roc_df.tpr,
+        label='{} (AUROC = {:.1%})'.format(label, metrics['auroc']))
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('Predicting RIT1 mutation from gene expression (ROC curves)')
+plt.legend(loc='lower right');
+
+
+# ## Investigate the predictions
+
+# In[24]:
+
+X_transformed = scale_post.transform(pca.transform(scale_pre.transform(X)))
```

Comment:
Did you want to put scale_pre, pca, and scale_post in a pipeline? This would really simplify things and help avoid errors.